kotlinx-coroutines-core/common/test/DurationToMillisTest.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2016-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package kotlinx.coroutines
+
+import kotlin.test.*
+import kotlin.time.*
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.nanoseconds
+import kotlin.time.Duration.Companion.seconds
+
+class DurationToMillisTest {
+
+    @Test fun negative_duration_coerced_to_zero_millis() = assertEquals(
+        expected = 0L,
+        actual = (-1).seconds.toDelayMillis(),
+    )
+
+    @Test fun zero_duration_coerced_to_zero_millis() = assertEquals(
+        expected = 0L,
+        actual = 0.seconds.toDelayMillis(),
+    )
+
+    @Test fun one_nanosecond_coerced_to_one_millisecond() = assertEquals(
+        expected = 1L,
+        actual = 1.nanoseconds.toDelayMillis(),
+    )
+
+    @Test fun one_second_coerced_to_1000_milliseconds() = assertEquals(
+        expected = 1_000L,
+        actual = 1.seconds.toDelayMillis(),
+    )
+
+    @Test fun mixed_component_duration_rounded_up_to_next_millisecond() = assertEquals(
+        expected = 999L,
+        actual = (998.milliseconds + 75909.nanoseconds).toDelayMillis(),
+    )
+
+    @Test fun one_extra_nanosecond_rounded_up_to_next_millisecond() = assertEquals(
+        expected = 999L,
+        actual = (998.milliseconds + 1.nanoseconds).toDelayMillis(),
+    )
+
+    @Test fun infinite_duration_coerced_to_Long_MAX_VALUE() = assertEquals(
+        expected = Long.MAX_VALUE,
+        actual = Duration.INFINITE.toDelayMillis(),
+    )
+
+    @Test fun negative_infinite_duration_coerced_to_zero() = assertEquals(
+        expected = 0L,
+        actual = (-Duration.INFINITE).toDelayMillis(),
+    )
+
+    @Test fun nanosecond_off_by_one_infinity_does_not_overflow() = assertEquals(
+        expected = Long.MAX_VALUE / 1_000_000,
+        actual = (Long.MAX_VALUE - 1L).nanoseconds.toDelayMillis(),
+    )
+
+    @Test fun millisecond_off_by_one_infinity_does_not_overflow() = assertEquals(
+        expected = Long.MAX_VALUE, // INFINITY duration.
+        actual = (Long.MAX_VALUE - 1L).milliseconds.toDelayMillis(),
+    )
```

Comment:
I actually don't think this is correct now that I look at it again. This duration has no nanosecond sub component and therefore shouldn't be rounded.